source/PlanetLabel.cpp

```diff
@@ -91,76 +81,139 @@ PlanetLabel::PlanetLabel(const Point &position, const StellarObject &object, con
+	if(innerAngle < 0.)
+	{
+		innerAngle = LINE_ANGLES[0];
+		SetBoundingBox(labelDimensions, innerAngle);
+	}
+
+	// Cache the offsets for both labels; center labels.
+	const Point offset = GetOffset(Angle(innerAngle).Unit(), labelDimensions) - labelDimensions * .5;
+	const double nameX = (labelDimensions.X() - bigFont.Width(name)) * .5;
+	nameOffset = Point(offset.X() + nameX, offset.Y() + BORDER);
+	const double governmentX = (labelDimensions.X() - font.Width(government)) * .5;
+	governmentOffset = Point(offset.X() + governmentX, nameOffset.Y() + nameHeight + 1.);
 }
 
 
 
-void PlanetLabel::Draw() const
+void PlanetLabel::Update(const Point &center, const double zoom)
 {
-	// Draw any active planet labels.
-	const Font &font = FontSet::Get(14);
-	const Font &bigFont = FontSet::Get(18);
+	position = (objectPosition - center) * zoom;
+	radius = objectRadius * zoom;
+}
 
-	// The angle of the outer ring should be reduced by just enough that the
-	// circumference is reduced by 6 pixels.
-	double innerAngle = LINE_ANGLE[direction];
-	double outerAngle = innerAngle - 360. * GAP / (2. * PI * radius);
-	Point unit = Angle(innerAngle).Unit();
-	RingShader::Draw(position, radius + INNER_SPACE, 2.3f, .9f, color, 0.f, innerAngle);
-	RingShader::Draw(position, radius + INNER_SPACE + GAP, 1.3f, .6f, color, 0.f, outerAngle);
 
-	if(!name.empty())
-	{
-		Point from = position + (radius + INNER_SPACE + LINE_GAP) * unit;
-		Point to = from + LINE_LENGTH * unit;
-		LineShader::Draw(from, to, 1.3f, color);
 
-		double nameX = to.X() + (direction < 2 ? 2. : -bigFont.Width(name) - 2.);
-		bigFont.DrawAliased(name, nameX, to.Y() - .5 * bigFont.Height(), color);
+void PlanetLabel::Draw() const
+{
+	// Don't draw if too far away from the center of the screen.
+	const double offset = position.Length() - radius;
+	if(offset >= 600.)
+		return;
+
+	// Fade label as we get farther from the center of the screen.
+	const Color labelColor = color.Additive(min(.5, .6 - offset * .001));
 
-		double governmentX = to.X() + (direction < 2 ? 4. : -font.Width(government) - 4.);
-		font.DrawAliased(government, governmentX, to.Y() + .5 * bigFont.Height() + 1., color);
-	}
+	// The angle of the outer ring should be reduced by just enough that the
+	// circumference is reduced by GAP pixels.
+	const double outerAngle = innerAngle - 360. * GAP / (2. * PI * radius);
+	RingShader::Draw(position, radius + INNER_SPACE, 2.3f, .9f, labelColor, 0.f, innerAngle);
+	RingShader::Draw(position, radius + INNER_SPACE + GAP, 1.3f, .6f, labelColor, 0.f, outerAngle);
+
+	const double barbRadius = radius + 25.;
 	Angle barbAngle(innerAngle + 36.);
 	for(int i = 0; i < hostility; ++i)
 	{
-		barbAngle += Angle(800. / (radius + 25.));
-		PointerShader::Draw(position, barbAngle.Unit(), 15.f, 15.f, radius + 25., color);
+		barbAngle += Angle(800. / barbRadius);
+		PointerShader::Draw(position, barbAngle.Unit(), 15.f, 15.f, barbRadius, labelColor);
 	}
+
+	// Draw planet label, if any.
```

Comment:
We've already drawn part of the planet label above, so specify that this is the name part of the planet label.
```suggestion
	// Draw planet name label, if any.
```